src/testing.py

```diff
@@ -89,6 +89,46 @@ def __call__(self, test_file):
       return Result(test=basename, success=False, output=e.output)
 
 
+def parse_exclude_files(fails, config_attributes):
+  ''' Parse the files containing tests to exclude (i.e. expected fails).
+  Each line may contain a comma-separated list of attributes restricting
+  the test configurations which are expected to fail. (e.g. JS engine
+  or optimization level). A test is only excluded if the configuration
+  has all the attributes specified in the exclude line. Lines which
+  have no attributes will match everything, and lines which specify only
+  one attribute (e.g. engine) will match all configurations with that
+  attribute (e.g. both opt levels with that engine). Returns a sorted list
+  of exclusions which match the attributes.
+  '''
+  excludes = {}  # maps name of excluded test to file from whence it came
+  if not config_attributes:
+    config_attributes = set()
+  for excludefile in fails:
+    f = open(excludefile)
+    for line in f:
+      line = line.strip()
+      if '#' in line:
+        line = line[:line.index('#')].strip()
+      if not line:
+        continue
+      tokens = line.split()
+      if len(tokens) > 1:
+        attributes = set(tokens[1].split(','))
+        if not attributes.issubset(config_attributes):
```

Comment:
I should have been more explicit that "I just now learned about set operators, and now let's all quietly forget about them" :)

But yeah, I like the symmetry of both objects being sets, if ever we want to do set-operations with `config_attributes`. It's really clear now that it's always a set, which cuts down lots of future code reading.